File: ruby/easy/paint_house.rb
# @param {Integer[][]} costs
# @return {Integer}
def min_cost(costs)
    return 0 unless !costs.empty?
    prev_red, prev_blue, prev_green = costs[0][0], costs[0][1], costs[0][2]
    for i in 1..costs.length-1
        red = [prev_blue, prev_green].min + costs[i][0]
        blue = [prev_red, prev_green].min + costs[i][1]
        green = [prev_red, prev_blue].min + costs[i][2]
        prev_red, prev_blue, prev_green = red, blue, green
    end
    return [prev_red, prev_blue, prev_green].min
end

File: javascript/easy/paintHouse.js
/**
 * @param {number[][]} costs
 * @return {number}
 */
var minCost = function(costs) {
    if(costs.length == 0) return 0;

    let min = Number.MAX_SAFE_INTEGER, memo = [[]];
    for(let i = 0; i < costs.length; i++) {
        memo[i] = [-1, -1, -1];
    }
    for(let i = 0; i < 3; i++) {
        min = Math.min(min, costs[0][i] + minimumCost(1, i, costs, memo));
    }
    return min;
};

var minimumCost = function(i, prev, costs, memo) {
    if(i == costs.length) {
        return 0;
    } else if(memo[i][prev] === -1) {
        let min = Number.MAX_SAFE_INTEGER;
        for(let j = 0; j < 3; j++) {
            if(j !== prev) {
                min = Math.min(min, (costs[i][j] + minimumCost(i+1, j, costs, memo)));
            }
        }
        memo[i][prev] = min;
    }
    return memo[i][prev];
}
